src/panels/lovelace/cards/hui-entities-card.ts

```diff
@@ -0,0 +1,191 @@
+    if (!this.config) {
+      return 0;
+    }
+    return this.config.title ? 1 : 0 + this.config.entities.length;
+  }
+
+  public setConfig(config: Config) {
+    const entities = processConfigEntities(config.entities);
+    for (const entity of entities) {
+      if (
+        entity.type === "call-service" &&
+        (!entity.service ||
+          !entity.name ||
+          !entity.icon ||
+          !entity.service_data ||
+          !entity.action_name)
+      ) {
+        throw new Error("Missing required property when type is call-service");
+      } else if (
+        entity.type === "weblink" &&
+        (!entity.name || !entity.icon || !entity.url)
+      ) {
+        throw new Error("Missing required property when type is weblink");
+      }
+    }
+
+    this.config = config;
+    this.configEntities = entities;
+
+    if (this.hass) {
+      this.requestUpdate();
+    }
+  }
+
+  protected render() {
+    if (!this.config || !this.hass) {
+      return html``;
+    }
+    const { show_header_toggle, title } = this.config;
+
+    const entities = this.configEntities!.map((conf) => conf.entity);
+
+    return html`
+      ${this.renderStyle()}
+      <ha-card>
+        ${
+          !title && !show_header_toggle
+            ? html``
+            : html`
+            <div class='header'>
+              ${!title ? html`` : html`<div class="name">${title}</div>`}
+              ${
+                show_header_toggle === false
+                  ? html``
+                  : html`
+                  <hui-entities-toggle 
+                    .hass="${this.hass}" 
+                    .entities="${entities}"
+                  >
+                  </hui-entities-toggle>`
+              }
+            </div>`
+        }
+        <div id="states">
+          ${repeat<EntityConfig>(
+            this.configEntities!,
+            (entityConf) => entityConf.entity,
+            (entityConf) => this.renderEntity(entityConf)
+          )}
+        </div>
+      </ha-card>
+    `;
+  }
+
+  private renderStyle() {
+    return html`
+      <style>
+        ha-card {
+          padding: 16px;
+        }
+        #states {
+          margin: -4px 0;
+        }
+        #states > div {
+          margin: 4px 0;
+        }
+        #states > div > * {
+          overflow: hidden;
+        }
+        .header {
+          @apply --paper-font-headline;
+          /* overwriting line-height +8 because entity-toggle can be 40px height,
+            compensating this with reduced padding */
+          line-height: 40px;
+          color: var(--primary-text-color);
+          padding: 4px 0 12px;
+          display: flex;
+          justify-content: space-between;
+        }
+        .header .name {
+          @apply --paper-font-common-nowrap;
+        }
+        .state-card-dialog {
+          cursor: pointer;
+        }
+      </style>
+    `;
+  }
+
+  private renderEntity(entityConf) {
+    const element = createRowElement(entityConf);
```

Comment:
this is not efficient. We're going to create new elements now on every re-render.

We should remove `hass` from the properties, and should manually find these elements in the DOM (`this.shadowRoot.querySelectorAll('.card > *')`) and update the hass property accordingly.

Reply:
We can detect when hass changes using a setter: `set hass(value) { … }`

Reply:
Im not sure I 100% know what to do here.

Are you saying render entity should only happen once? How can we flag as the first time?

Then using `set hass` we should do something like this?

```
set hass(hass) {
    this.shadowRoot!.querySelectorAll('.card > *').forEach(element => {
      element.hass = hass;
    });
  }
```

Reply:
Actually, yeah, this code is fine. However, remove hass from properties and add the method you described for `set hass`

Reply:
When taking the `element.hass = hass` and adding it to set hass. The entities come back as unavailable rows. I believe this is because when set hass runs the first time `this.shadowRoot.querySelector("ha-card")` doesnt exists so it is unable to set it before the elements call for hass.

Reply:
Ah, in that case let's make sure we set `hass` inside this method too.

Reply:
renderEntity